### ISSUE TITLE
one.list keywords and plurals

### ISSUE THREAD

**Issue by kdharris101** [int-brain-lab/ibllib#17]
For one.list it seems the correct syntax is: 

`one.list(None, keyword='subjects')`
but also
`one.list(eids, keyword='subject')`

This is confusing!! 

I also can't do `one.list(eids, keyword='dataset_types')`, even with any combination of plurals I can find. 

Aren't the keywords supposed to be the same as the search terms? In which case keyword could have any of the values ('dataset_types', 'users', 'subjects', 'date_range')? 

I realize date_range could present a problem (since it is a range) but other than that can they not be the same?

Also, could keywords=None return a list for all keywords?

**Comment by kdharris101** [comment]
Actually how about this (to get around the problem of lists of lists):

`one.list(eids, search_terms)` returns a dict. There is one key for each entry in the list `search_terms`, whose value contains a list of the same length as `eids`. 

If you pass `search_terms=None`, it will default to all the search terms there are. If you pass `eids=None`, it will default to all the eids there are. The default argument for `search_terms`  is `'dataset_types'`, but that can also be passed as part of a longer list of search terms, and is also included in the None input option.

**Comment by oliche** [comment]
### First Post
The problem is that the keyword refers to 3 different things depending on the use:
in `one.list(eids, keyword = 'subject')`, it refers to the field subject of the sessions referenced by eids
in `one.list(None, keyword = 'subjects')` it refers to the table subjects (endpoint)
in `one.search(subject='toto')` subject refers to a filter object in the Django API, that overlaps by default with the fields, but may be customized (as in `date_range`).

To address this, the one class opens with the following
```
ENDPOINT_LIST = ('dataset-types', 'users', 'subjects')
SESSION_FIELDS = ('subject', 'users', 'lab', 'type', 'start_time', 'end_time')
SEARCH_TERMS = ('dataset_types', 'users', 'subjects', 'date_range')
```
To help the users, I've tried to make meaningful error messages that list possible values when fields doesn't exist, but I still wanted to have keywords matching the names of the objects they referred to. 

### Suggestion
Thinking about user-friendliness,  we may not be  leveraging auto-completion as much as we could. We could be using classes/dataclasses/named tuples as such:
`one.list.dataset_types(), one.list.any_endpoint()`
`one.list.dateset_types(eid), one.list.subjects(eid)`
This makes the functions more usable, less prone to nagging spelling issues, and the ONE standard can stay flexible on the methods/fields to implement with a few must-haves.

This will also make the code consistent with the Matlab structures, easy to switch.

**Comment by kdharris101** [comment]
> We could be using classes/dataclasses/named tuples as such `one.list.any_endpoint()`

Very nice idea! And I guess this could even be done dynamically? I.e. if in the future the list of search terms is found from the database, the one object could be given these functions dynamically - at least in Python?

**Comment by oliche** [comment]
### List
Yes I was wondering if I had to do the set of unique values or not. One user case would be to get all information for each session in the same order as the query for further selection, another use-case is to get the unique set.

I implemented the `'details'` to get all information about each queried session and will probably keep an option like that, in the same order as the session. 

### Search
> Does this make sense? 

Yes ! The good thing is that we could re-use most of Django operators this way, and overload or create the one we need. I need to have a look at Datajoint query syntax.
What may be missing above is the relation between queries (AND (default) / OR) but this is probably for further down the road.

### Keywords flexibility
There is one thing I am sure of: if we stay with named arguments (no autocomplete), the translation of input arguments will be implemented using dictionaries, where the key name is the user possible input keywords and the values the actual implementation:

```
_ENDPOINTS = {
     'data': 'dataset-types',
     'dataset': 'dataset-types',
     'datasets': 'dataset-types',
     'dataset-types': 'dataset-types',
     'users': 'users',
     'user': 'users',
     'subject': 'subjects',
     'subjects': 'subjects'
}
```
In short we design the functions so that typos from the users are allowed.

**Comment by rossant** [comment]
For search, why not just use the django syntax ? https://docs.djangoproject.com/en/2.1/ref/models/querysets/#field-lookups

**Comment by oliche** [comment]
What I'm gathering from all the discussions above is that the implementation and the user syntax are bound to be different. 

The keywords do not match the database field names (plurals), one argument refers to different objects kinds (tables,fields or search filter keywords).

From the user perspective, I think we shouldn't assume that the user will have to know that x field is in another table (double underscore) nor the exact name of searchable fields: this pretty much rules out the Django syntax. Plus some operators we'll have to implement with Q functions.
For our implementation, yes !

**Comment by rossant** [comment]
We don't have to follow the django syntax to the letter. But why inventing a new syntax that does the same thing, when we can just use the django syntax?

We don't have to care about which fields belong to which table, it's just a matter of using the appropriate keywords and operators with `__`. This syntax doesn't have to be strictly the one we'd use in the django console, it can just be loosely inspired by it. As a user I think I'd prefer to type something like `date__gt='20170101'` rather than `date=('>', 'Jan 1 2017')`.

For the dataset types, we can come up with anything we'd like, e.g. `dataset_types=['spikes.times', 'spikes.clusters', 'clusters.probes']`.

**Comment by kdharris101** [comment]
Sounds good. To be clear, you are suggesting:

1.       Rather than passing relationships as tuples, using the Django `field__relation` when a relationship needs to be specified.

2.       Not passing Django queries straight to Django, but just copying some of their syntax

3.       Not using this syntax to link between tables – but instead having linked items (such as dataset_types) be translated to a list, and defining new relationship operators for these lists, also encoded in field__relation form?

**Comment by rossant** [comment]
@kdharris101 yes, absolutely!

**Comment by kdharris101** [comment]
OK sounds good! Olivier what do you think?

**Comment by kdharris101** [comment]
If we can use Datajoint syntax that would be great. It will be familiar to more users than Django.

Let’s think about some specific use cases. Again, the most important one is to find all experiments that have all in a list of DatasetTypes. I’m not sure how to do that in DataJoint. I guess we could ask them!

A more difficult one is to find all recordings that have at least 8 narrow-spiking neurons recorded in area CA1. I can’t think of any way to do this easily, including in the system I just suggested.

**Comment by kdharris101** [comment]
OK here's a though of how to use the same framework to do searches for experiments with enough cells of a certain type.

Each experiment has an attribute called `cells` or something like that, which is a data frame - i.e. a table containing an entry for each cell recorded in that experiment, with multiple fields for each cell like brain region, mean firing rate, isolation quality, putative cell type, etc.

We then define an operator that tests if the experiment has at least N cells matching a certain combination of criteria. Something like:

`one.search(cellCounts__gt(10,brainLocation='CA1', spikeWidth='narrow', firingRate__gt=5))`

The idea here is that cellCounts__gt(N, conditions...) would specifying a condition that there must be at least N cells obeying the specified conditions. In this case there would need to be at least 10 cells with location in CA1, narrow spike width, firing rate at least 5Hz.

Not sure if this syntax is legal in python though... but something along these lines...

**Comment by kdharris101** [comment]
Yes, that’s a good point. There’s no need to spend too much time implementing complicated search features in ONE – unless we find users actually want them because they are not happy with DJ for some reason.

**Comment by rossant** [comment]
closing the issue here and continuing the discussion on the ONE v2 proposal
